### PR TITLE
Add durability

### PR DESCRIPTION
**Pull request by whilo**
This PR adds durability to the persistent-sorted-set.

### CODE REVIEW

**Comment by tonsky** [comment]
Thanks! Looks dope, I’ll take a detailed look next week

**Comment by whilo** [comment]
I should have marked this as a draft PR somehow. We would be open to maintain a fork if needed, but it is preferable for us to keep our stacks shared with you since so far we have not ventured far away from the DataScript Datalog approach. Btw. this also applies to the Datalog parser we have factored out and could be extended to other bits of the stack as well. Building databases is a lot of work. 

This is still a first prototype of how to add durability, at the moment I opted to batch load all child nodes at once (which is different to common node-wise loading strategies as employed in the hitchhiker-tree). I would actually like to leave full flexibility to the loading strategy, so probably we should allow the loader to decide which children to load and when. For a full parametrization of loading strategies it would also make sense to add some fields to nodes containing access statistics, loading times etc. We have come to the conclusion to port Datahike onto a durable version of the persistent-sorted-set because it has the fastest iterators (5x times faster or so than the hitchhiker-tree in simple benchmarks) and we have run into performance bumps with edge cases in our lazy upsert operations in the hitchhiker-tree.

Nonetheless it will be interesting in the future to also think about readding the messaging overlay of the hitchhiker-tree onto the persistent set (I have an idea to do it better than the current hh-tree). I am happy to also have a call about this at some point if you are interested. At this point your input will definitely be helpful to steer this in a direction that is also of interest to you. A particular design objective is to not regress on in-memory performance compared to your current implementation, so your input now will definitely be helpful in ensuring that. I had to remove the `final` of `_children` for lazy loading for instance and I also still need to check whether synchronization/locking is needed around node updates during loading.

And as always, I deeply appreciate your care in this implementation, it is a pleasure to work on your codebases!

File: src-java/me/tonsky/persistent_sorted_set/Node.java [hunk 1]
```diff
@@ -326,8 +346,15 @@ public String str(int lvl) {
       sb.append("\n");
       for (int j=0; j < lvl; ++j)
         sb.append("| ");
-      sb.append(_keys[i] + ": " + _children[i].str(lvl+1));
+      ensureChildren();
```
Comment (tonsky) [inline src-java/me/tonsky/persistent_sorted_set/Node.java:419]: Not needed here?

File: src-clojure/me/tonsky/persistent_sorted_set.clj [hunk 2]
```diff
@@ -75,41 +75,138 @@
...
+      (let [loader (.-_loader n)
+            children (into-array Leaf (map #(-flush %) (.-_children n)))
+            address (.store loader children)]
+        (set! (.-_address n) address)
+        (set! (.-_isDirty n) false)
+        (set! (.-_isLoaded n) false)
+        (set! (.-_children n) nil)
+        n))))
+
+(defprotocol NodeToMap
```
Comment (tonsky) [inline src-clojure/me/tonsky/persistent_sorted_set.clj:151]: This doesn’t have to be in the lib? Only needed for example?

File: project.clj [hunk 3]
```diff
@@ -6,6 +6,7 @@
   :dependencies [
     [org.clojure/clojure       "1.10.0"   :scope "provided"]
     [org.clojure/clojurescript "1.10.520" :scope "provided"]
+    [io.replikativ/konserve    "0.6.0-alpha3"]
```
Comment (tonsky) [inline project.clj:9]: Remove?

File: project.clj [hunk 4]
```diff
@@ -17,7 +18,7 @@
   :java-source-paths ["src-java"]
   :test-paths        ["test-clojure"]
 
-  :javac-options ["-target" "8" "-source" "8" "-bootclasspath" ~(str (or (System/getenv "JAVA8_HOME") (throw (Exception. "Please set JAVA8_HOME"))) "/jre/lib/rt.jar")]
+  ;:javac-options ["-target" "8" "-source" "8" "-bootclasspath" ~(str (or (System/getenv "JAVA8_HOME") (throw (Exception. "Please set JAVA8_HOME"))) "/jre/lib/rt.jar")]
```
Comment (tonsky) [inline project.clj:21]: Restore? Unless needed for some reason

File: src-clojure/me/tonsky/persistent_sorted_set.clj [hunk 5]
```diff
@@ -75,41 +75,138 @@
...
+  ([loader] (PersistentSortedSet/EMPTY))
+  ([loader & keys] (from-sequential compare keys loader)))
+
+
+;; code for durability
+
+(defprotocol FlushNode
+  (-flush [n]))
+
+(extend-protocol FlushNode
```
Comment (tonsky) [inline src-clojure/me/tonsky/persistent_sorted_set.clj:134]: Can we move it to Leaf/Node? Does it have to be a separate protocol?

File: src-java/me/tonsky/persistent_sorted_set/Loader.java [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package me.tonsky.persistent_sorted_set;
+import java.util.UUID;
+
+public interface Loader {
```
Comment (tonsky) [inline src-java/me/tonsky/persistent_sorted_set/Loader.java]: Since it does both load and save, maybe rename to Storage?

**Review by tonsky — COMMENTED** [review]
I’m happy to move in this direction. Let’s discuss it on a call

I’ve made a few notes just to remember. We’ll probably need a CLJS version too

**Comment by tonsky** [comment]
Topics to discuss: grouping leaves before save. 1 file per node is super inefficient

**Comment by tonsky** [comment]
Allow null for Loader

File: src-clojure/me/tonsky/persistent_sorted_set.clj (hunk 5 above)
Comment (whilo) [reply src-clojure/me/tonsky/persistent_sorted_set.clj:134]: We can, although it is in fact a matter for how you do the writing and does not even have to be in the repo.

File: src-java/me/tonsky/persistent_sorted_set/Loader.java (hunk 6 above)
Comment (whilo) [reply src-java/me/tonsky/persistent_sorted_set/Loader.java]: Done.

**Comment by whilo** [comment]
Is it ok to have a dummy `NullLoader` that does nothing and is the default argument for `sorted-set` or do you want to be able to literally pass in `null` as the argument (and where specifically can things be `null`)?

File: src-clojure/me/tonsky/persistent_sorted_set.clj (hunk 5 above)
Comment (whilo) [reply src-clojure/me/tonsky/persistent_sorted_set.clj:134]: I guess it would make sense to have it there and move it into Leaf/Node.

**Comment by tonsky** [comment]
NullLoader should be fine, hope it doesn’t bring perf regressions

**Comment by whilo** [comment]
Cool, sounds good :+1: .

**Comment by whilo** [comment]
I have added the null-loader and have done a sizeable update covering thread safety (hopefully) under cache removals. I have stress tested this with Datahike benchmarks, but it might still have some issues, so feedback is welcome. You might not like this since it increases coordination overhead and potentially reduces the performance a bit, but there is a way to recover the original unlocked performance by dispatching on the null-loader presence and using a dummy locking mechanism in this case.

**Comment by tonsky** [comment]
Ok I’ll take a look

**Comment by whilo** [comment]
I also added a mark phase for garbage collection. I can move this to another namespace ofc., feel free to make suggestions. This is the complementary functionality to https://github.com/replikativ/konserve/blob/main/src/konserve/gc.cljc#L7.

**Comment by tonsky** [comment]
I liked your approach, but there are too many details to discuss and change. E.g.:

- Extra code that needs to be removed (map->node, NodeToMap, FlushNode, mark, StorageBackend/hitCache & store)
- Contracts were broken in some places (from-sequential/sorted-set/sorted-set-by, PersistentSortedSet/EMPTY)

Also, I wanted to play with implementation myself:

- ReentrantReadWriteLock seems excessive, wouldn’t ReentrantLock be enough?
- Happy path should contain no locks at all
- Allow Address to be any opaque Object, not just UUID
- Rigorous tests and benchmarks, esp. for new lazy-loaded set
- Experiment with SoftReference to allow JVM to unload segments on GC pressure

I figured that it’ll be faster to address all that in my own branch instead of communicating and requesting every change in PR.

The branch is at https://github.com/tonsky/persistent-sorted-set/tree/storage. It’s not finalized yet, but I want to give you a heads up.

Please take a look and let me know if something like that would work for you? Clojure tests & serialization example are in https://github.com/tonsky/persistent-sorted-set/blob/storage/test-clojure/me/tonsky/persistent_sorted_set/test_storage.clj

The branch also has a refactoring that unified Leaf and Node and gives them a little bit nicer API regarding accessing internal fields. I figured that is necessary since serialization code would work with internal representation of the nodes.

Anyways, let me know what you think. I’ll try to finalize it this week

**Comment by tonsky** [comment]
Pushed 0.2.0! Thanks and congrats!

**Pull request closed without merging**